custom/onse/tasks.py

```diff
@@ -0,0 +1,298 @@
+    queue='background_queue',
+)
+def update_facility_cases_from_dhis2_data_elements():
+    try:
+        dhis2_server = ConnectionSettings.objects.get(
+            domain=DOMAIN, name=CONNECTION_SETTINGS_NAME
+        )
+    except ConnectionSettings.DoesNotExist:
+        _soft_assert(False, (
+            f'ConnectionSettings {CONNECTION_SETTINGS_NAME!r} not found in '
+            f'domain {DOMAIN!r} for importing DHIS2 data elements.'))
+        return
+
+    facility_case_blocks = get_case_blocks()
+    facility_case_blocks = set_case_updates(dhis2_server, facility_case_blocks)
+    save_cases(facility_case_blocks)
+
+
+def get_case_blocks() -> Iterable[CaseBlock]:
+    case_accessors = CaseAccessors(DOMAIN)
+    for case_id in case_accessors.get_case_ids_in_domain(type=CASE_TYPE):
+        case = case_accessors.get_case(case_id)
+        if not case.external_id:
+            # This case is not mapped to a facility in DHIS2.
+            continue
+        case_block = CaseBlock(
+            case_id=case.case_id,
+            external_id=case.external_id,
+            case_type=CASE_TYPE,
+            case_name=case.name,
+            update={},
+        )
+        yield case_block
+
+
+def set_case_updates(
+    connection_settings: ConnectionSettings,
+    case_blocks: Iterable[CaseBlock]
+) -> Iterable[CaseBlock]:
+    """
+    Fetch all the data elements for the data set, because we don't know
+    in advance what category option combos to query for
+    """
+    # e.g. https://play.dhis2.org/dev/api/dataValueSets?orgUnit=jNb63DIHuwU
+    #   &period=2020Q2
+    #   &dataSet=QX4ZTUbOt3a  <-- The data set of the data element
+    # not https://play.dhis2.org/dev/api/dataValues?ou=jNb63DIHuwU
+    #   &pe=2020Q2
+    #   &de=gQNFkFkObU8  <-- The data element we want
+    #   &co=L4P9VSgHkF6  <-- We don't know this, and without it we get a 409
+    requests = connection_settings.get_requests()
+    endpoint = '/api/dataValueSets'
+    params = {'period': get_last_quarter()}
+    for case_block in case_blocks:
+        params['orgUnit'] = case_block.external_id
+        data_set_cache = {}
+        for case_property, (data_element, data_set) in CASE_PROPERTY_MAP.items():
+            if data_set not in data_set_cache:
+                params['dataSet'] = data_set
+                response_json = requests.get(endpoint, params).json()
+                data_set_cache[data_set] = response_json.get('dataValues', None)
+
+            if data_set_cache[data_set] is None:
+                # No data for this facility. `None` = "We don't know"
+                case_block.update[case_property] = None
+            else:
+                value = 0
+                for data_value in data_set_cache[data_set]:
+                    if data_value['dataElement'] == data_element:
+                        value += int(data_value['value'])
+                case_block.update[case_property] = value
+        yield case_block
```

Comment:
nit: this function is a little on the complex side.  With a small class, you could separate out the body of the loop and extract the data_set caching business so the case block updates are clearer.  I have no idea what most of these variables refer to, so it'd be nice to see it at a higher level of abstraction.